src/pipeline/add_pipeline.rs

```diff
@@ -0,0 +1,99 @@
+use crate::field::Fp31;
+use crate::pipeline::PipelineExt;
+use crate::shamir::{LagrangePolynomial, SecretSharing, Share};
+use futures::stream::StreamExt;
+use futures::Stream;
+use std::num::NonZeroU8;
+
+pub struct Leader {
+    k: NonZeroU8,
+}
+
+impl Leader {
+    #[allow(dead_code)]
+    pub fn run(
+        &self,
+        source: impl Stream<Item = (Share<Fp31>, Share<Fp31>)>,
+        h1_res: impl Stream<Item = Share<Fp31>>,
+        h2_res: impl Stream<Item = Share<Fp31>>,
+    ) -> impl Stream<Item = Fp31> {
+        let k = self.k;
+        source
+            .map(|(x_share, y_share)| &x_share + &y_share)
+            .zip3(h1_res, h2_res)
+            .map(move |(source_share, h1_share, h2_share)| {
+                SecretSharing::reconstruct(
+                    &[source_share, h1_share, h2_share],
+                    &LagrangePolynomial::new(k).unwrap(),
+                )
+                .unwrap()
+            })
+    }
+}
+
+pub struct Helper {}
+
+impl Helper {
+    #[allow(clippy::unused_self)] // may need to use it eventually
+    pub fn run(
+        &self,
+        source: impl Stream<Item = (Share<Fp31>, Share<Fp31>)>,
+    ) -> impl Stream<Item = Share<Fp31>> {
+        source.map(|(share_x, share_y)| &share_x + &share_y)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::field::Field;
+    use futures::stream::StreamExt;
+    use rand::rngs::mock::StepRng;
+    use rand::{thread_rng, Rng};
+
+    fn gen_numbers(len: usize) -> impl Stream<Item = (NonZeroU8, NonZeroU8)> {
+        let mut rng = StepRng::new(1, 1);
+
+        futures::stream::iter(std::iter::from_fn(move || {
+            let x = NonZeroU8::new(rng.gen_range(1..Fp31::PRIME)).unwrap();
+            let y = NonZeroU8::new(rng.gen_range(1..Fp31::PRIME)).unwrap();
+            Some((x, y))
+        }))
+        .take(len)
+    }
+
+    #[tokio::test]
+    async fn add_pipeline() {
+        let nums_iter = gen_numbers(10);
+        let k = NonZeroU8::new(3).unwrap();
+        let n = NonZeroU8::new(4).unwrap();
+
+        let mut unzipped = nums_iter
+            .map(move |(x, y)| {
+                let shamir = SecretSharing::new(k, n).unwrap();
+                let x_shares = shamir.split(Fp31::from(x.get()), thread_rng());
+                let y_shares = shamir.split(Fp31::from(y.get()), thread_rng());
+                let share0 = (x_shares[0].clone(), y_shares[0].clone());
+                let share1 = (x_shares[1].clone(), y_shares[1].clone());
+                let share2 = (x_shares[2].clone(), y_shares[2].clone());
+                (share0, share1, share2)
+            })
+            .unzip3();
+        let (shares_1, shares_2, shares_3) = unzipped.output();
```

Comment:
I am not sure if you actually need this zip/unzip - could it just be `stream::iter` over each helper's portion of shares?

Reply:
yes, that is possible. I decided to implement the streaming version of zip/unzip because i thought we may actually need that going forward, and wanted to see what that might look like